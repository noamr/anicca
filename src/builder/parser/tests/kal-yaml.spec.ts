import {parse} from 'yaml'
const {parseKal} = require('../index')

const kalYaml = `
view myView:
    '#output':
        content: value
    '#increment':
        on click:
            dispatch increment to myController

controller myController:
    state root:
        entering:
            value = 0

        on increment:
            value += 1

let value: u32
`

describe('kal yaml parsing', () => {
    it('should be valid yaml', () => {
        expect(parse(kalYaml)).toMatchSnapshot()
    })
    it('should be parsed to json', () => {
        expect(parseKal(parse(kalYaml))).toMatchSnapshot()
    })
    it('should fail on internals', () => {
        expect(() => parseKal(parse(`
            let @internal: u32
        `))).toThrow()
})
    it('should fail on internal in states', () => {
        expect(() => parseKal(parse(`
        controller myController:
            state @root:
                entering:
                    value = 0
            `))).toThrow()
    })
    it('should allow internal in states in internal mode', () => {
        expect(() => parseKal(parse(`
        controller myController:
            state @root:
                entering:
                    value = 0
            `), {internal: true})).not.toThrow()
    })
    it('should accept internals in internal mode', () => {
        expect(() => parseKal(parse(`
            let @internal: u32
        `), {internal: true})).not.toThrow()
    })
})
