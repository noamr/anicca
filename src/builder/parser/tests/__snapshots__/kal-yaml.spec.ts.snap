// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kal yaml parsing should be parsed to json 1`] = `
Array [
  Object {
    "name": "myView",
    "rules": Array [
      Object {
        "declarations": Array [
          Object {
            "src": Object {
              "$ref": "value",
            },
            "targetType": "content",
            "type": "Bind",
          },
        ],
        "selector": "#output",
      },
      Object {
        "declarations": Array [
          Object {
            "actions": Array [
              Object {
                "event": "increment",
                "target": "myController",
                "type": "Dispatch",
              },
            ],
            "eventType": "click",
            "type": "DOMEvent",
          },
        ],
        "selector": "#increment",
      },
    ],
    "type": "View",
  },
  Object {
    "name": "myController",
    "rootState": Object {
      "children": Array [
        Object {
          "actions": Array [
            Object {
              "src": Object {
                "$primitive": 0,
              },
              "target": Object {
                "$ref": "value",
              },
              "type": "Assign",
            },
          ],
          "type": "OnEntry",
        },
        Object {
          "actions": Array [
            Object {
              "source": Object {
                "args": Array [
                  Object {
                    "$ref": "value",
                  },
                  Object {
                    "$primitive": 1,
                  },
                ],
                "op": "plus",
              },
              "target": Object {
                "$ref": "value",
              },
              "type": "Assign",
            },
          ],
          "event": "increment",
          "type": "Transition",
        },
      ],
      "default": undefined,
      "name": "root",
      "type": "State",
    },
    "type": "Controller",
  },
  Object {
    "name": "value",
    "type": "Let",
    "valueType": "u32",
  },
]
`;

exports[`kal yaml parsing should be valid yaml 1`] = `
Object {
  "controller myController": Object {
    "state root": Object {
      "entering": "value = 0",
      "on increment": "value += 1",
    },
  },
  "let value": "u32",
  "view myView": Object {
    "#increment": Object {
      "on click": "dispatch increment to myController",
    },
    "#output": Object {
      "content": "value",
    },
  },
}
`;
